debezium-server-milvus/src/main/java/io/debezium/server/milvus/MilvusChangeConsumer.java

```diff
@@ -0,0 +1,207 @@
+
+        for (final ChangeEvent<Object, Object> record : records) {
+            LOGGER.trace("Received event '{}'", record);
+
+            final var sourceRecord = toSourceRecord(record);
+
+            // Milvus does not support dots in collection names so we by default replace them
+            // with underscores so the user does not need to provide router or mapper
+            final var collectionName = streamNameMapper.map(record.destination()).replace('.', '_');
+
+            if (isSchemaChange(sourceRecord)) {
+                LOGGER.debug("Schema change event, ignoring it");
+                committer.markProcessed(record);
+                continue;
+            }
+
+            if (record.key() == null) {
+                throw new DebeziumException("Milvus does not support collections without primary key");
+            }
+            schema.validateKey(collectionName, sourceRecord.keySchema());
+
+            if (sourceRecord.value() == null) {
+                deleteRecord(collectionName, record, committer);
+            }
+            else if (Envelope.isEnvelopeSchema(sourceRecord.valueSchema())) {
+                final var valueStruct = (Struct) sourceRecord.value();
+                switch (Operation.forCode(valueStruct.getString(Envelope.FieldName.OPERATION))) {
+                    case Operation.READ:
+                    case Operation.CREATE:
+                    case Operation.UPDATE:
+                        upsertRecord(collectionName, record, committer);
+                        break;
+                    case Operation.DELETE:
+                        deleteRecord(collectionName, record, committer);
+                        break;
+                    default:
+                        LOGGER.info("Unsupported operation, skipping record '{}'", record);
+                }
+            }
+            else {
+                // Extracted new record state
+                upsertRecord(collectionName, record, committer);
+            }
+        }
+
+        committer.markBatchFinished();
+    }
+
+    protected SourceRecord toSourceRecord(final ChangeEvent<Object, Object> record) {
+        // Milvus sink requires access to the message schema so it can do schema evolution
+        // This is not a part of public Debezium Engine API but is an implementation detail on
+        // which Debezium Server can rely
+        @SuppressWarnings("rawtypes")
+        final var sourceRecord = ((EmbeddedEngineChangeEvent) record).sourceRecord();
+        return sourceRecord;
+    }
+
+    private void upsertRecord(String collectionName, ChangeEvent<Object, Object> record,
+                              RecordCommitter<ChangeEvent<Object, Object>> committer)
+            throws InterruptedException {
+        final var data = getValue(record, toSourceRecord(record));
+
+        final var request = UpsertReq.builder()
+                .collectionName(collectionName)
+                .data(List.of(data))
```

Comment:
In the future we can possibly add caching of the records from given batch and do the bulk update.

Reply:
I agree for starter I just intend to keep things simple and stupid.